### DISCUSSION TITLE
how to add variables of process.env in nuxt3  ?

### DISCUSSION THREAD

**Discussion by sea-lanya** [nuxt/nuxt#15931] [Questions]
### Environment

nuxt3 ^3.0.0-27243104.5e903ae
mac

### Describe the bug

I try all the methods but not success

### Reproduction

I find some solutions like https://github.com/nuxt/framework/pull/1086 but not success

**Comment by lukaszflorczak** [comment]
I don't see any difference between Nuxt 2. 

```ts
publicRuntimeConfig: {
    MY_VARIABLE: process.env.MY_VARIABLE
}
```

Access to the value is by `$config.MY_VARIABLE`. I use `3.0.0-27242646.117b3aa` now and it works. 

EDIT: I updated to `3.0.0-27243104.5e903ae` and also works.

**Comment by wobsoriano** [comment]
in your nuxt config

```ts
export default defineNuxtConfig({
  runtimeConfig: {
    secret: process.env.SECRET,
    public: {
      whatever: process.env.YOUR_WHATEVER
    }
  }
}
```

Then in components

```html
<script setup lang="ts">
  const config = useRuntimeConfig()
  config.public.whatever
</script>
```

**Reply by vedranmviu** [reply]
How to do this outside of setup?

**Reply by ReaganM02** [reply]
@vedranmviu ,  You can use the `useRuntimeConfig()` outside script setup in two ways. 

- `const config = useRuntimeConfig()`
`config.JWT_SECRET`

- `useRuntimeConfig().JWT_SECRET` - One line code

`~~nux.config.js`

```
runtimeConfig: {
 JWT_SECRET: process.env.JWT_SECRET,
},
```

**Comment by sea-lanya** [comment]
thank @wobsoriano   @lukaszflorczak 

but my code include some others sdk 

sdk code require process.env.xxxxxx settings

so  I want to konw how to add variables like   process.env.SECRET

if I use  like this 
```
const config = useRuntimeConfig()
config.secret
```

I must change sdks code .... But I  can't do this.

**Comment by wobsoriano** [comment]
> thank @wobsoriano @lukaszflorczak
> 
> but my code include some others sdk
> 
> sdk code require process.env.xxxxxx settings
> 
> so I want to konw how to add variables like process.env.SECRET
> 
> if I use like this
> 
> ```
> const config = useRuntimeConfig()
> config.secret
> ```
> 
> I must change sdks code .... But I can't do this.

Do you have a `.env` file?

**Reply by sea-lanya** [reply]
I try it, but not sucess

![image](https://user-images.githubusercontent.com/9654331/137906446-35b1e38d-ac69-4fd8-bcb5-73ac8f16e37e.png)

```
beforeCreate() {
	console.log("env:", process.env)
  },
```
show 

```
{
BASE_URL: "/_nuxt/",
DEV: true,
MODE: "development",
PROD: false,
SSR: false
}

```

I also tried  .env.development

**Reply by wobsoriano** [reply]
Hmm not sure about Options API. I'm using Composition API to access it

```html
<script setup lang="ts">
  const config = useRuntimeConfig()
  console.log(config.secret)
</script>

<template>
  <div>Hello World</div>
</template>
```

**Reply by sea-lanya** [reply]
I wonder if vite caused it.

**Reply by wobsoriano** [reply]
@pi0 how would someone access runtime config if they're using OAPI?

**Reply by pi0** [reply]
Would you please elaborate more? I might help to complete an example if making sandbox

**Reply by wobsoriano** [reply]
He's using Options API and can't access the env variables

**Reply by sea-lanya** [reply]
https://github.com/mlsjla/nuxt3-starter
@pi0   This is a demo

I want to add variables in  process.env and get process.env.variables in other files like 'app.vue'

```
<script>
export default {
  setup() {
    onMounted(() => {
      // I want to get process.env here ...
      console.log('I cat get process.env here ...')
      // I can get process.env. But I don't know how to add variables in process.env
      console.log(process.env)
      // Like this
      console.log(process.env.variables)

      /*
       *  But I want to add other variables in process.env
       *  Like process.env.Base_URL or process.env.TIMEOUT etc..
       *  Because I have some sdks require process.env.variables
       *  
       *  If I use process.env with 'const config = useRuntimeConfig() ' , I must edit my sdks code
       *  But I can't
       */

    })
  },
}
</script>

```

**Reply by wobsoriano** [reply]
Then you really have to update your SDK

**Reply by sea-lanya** [reply]
thank u very much...

**Reply by milesingrams** [reply]
Im having issues with this as well and as far as I can tell the most recent version of Firebase Realtime Database is not possible to use because of the inability to set process.env.FIREBASE_DATABASE_EMULATOR_HOST

In the firebase initialization code the firebase web package tries to run this:

```javascript
let dbEmulatorHost: string | undefined = undefined;
  if (typeof process !== 'undefined') {
    dbEmulatorHost = process.env[FIREBASE_DATABASE_EMULATOR_HOST_VAR];
  }
```

In Nuxt 3 since I cannot set that env variable in any way it is impossible to get the Realtime Database up and running. Truthfully this seems like a bad design on the Firebase side to rely on process.env so heavily, but I would love if there's a workaround to set process.env variables both client and server side.

**Comment by pi0** [comment]
Please use discussions for questions

**Comment by mario-neuhold** [comment]
Regardless of his SDK, the suggested methods only work with ssr. I couldn't get it working with ssr: false (config = empty). Same when using nuxt2+bridge.
Am I missing something?

**Reply by wobsoriano** [reply]
Yep can confirm this.

with `ssr: true`, `publicRuntimeConfig` and `privateRuntimeConfig` works

```html
<script setup lang="ts">
const config = useRuntimeConfig()
console.log(config)
</script>
```

with `ssr: false`,  `publicRuntimeConfig` or `privateRuntimeConfig` is an empty object `{}`.

To get values from `.env` when `ssr: false`, [prefix the variables with VITE_](https://vitejs.dev/guide/env-and-mode.html#env-files). 

```
VITE_WEBSITE_URL=https://mywebsite.co
```

```html
<script setup lang="ts">
console.log(import.meta.env.VITE_WEBSITE_URL)
</script>
```

No need to add them in your `nuxt.config.ts`.

FYI: This is without bridge

**Reply by mario-neuhold** [reply]
Thanks for confirming and providing that workaround.
Is there a way to get the values without vite and bridge? (old nuxt2 project)

**Reply by Shohzod-Abdusamatov-7777777** [reply]
Thanks

**Comment by milesingrams** [comment]
Im having issues with this as well and as far as I can tell the most recent version of Firebase Realtime Database is not possible to use because of the inability to set process.env.FIREBASE_DATABASE_EMULATOR_HOST

In the firebase initialization code the firebase web package tries to run this:

```javascript
let dbEmulatorHost: string | undefined = undefined;
  if (typeof process !== 'undefined') {
    dbEmulatorHost = process.env[FIREBASE_DATABASE_EMULATOR_HOST_VAR];
  }
```

Resulting in this error:

<img width="713" alt="Screen Shot 2021-11-16 at 2 35 57 PM" src="https://user-images.githubusercontent.com/4930881/142053566-0e3edb0c-1821-4c52-99fc-83475ae69880.png">

In Nuxt 3 since I cannot set that env variable in any way it is impossible to get the Realtime Database up and running. Truthfully this seems like a bad design on the Firebase side to rely on process.env so heavily, but I would love if there's a workaround to set process.env variables both client and server side.

**Reply by gavin-hardaker** [reply]
If you're still having this issue, I made a small work around, where you're starting your firebase instance throw this in:

```javascript
import { useRuntimeConfig } from '#app'
const config = useRuntimeConfig()
process.env = config
initializeApp({
      apiKey: config.firebase.apiKey,
      authDomain: config.firebase.authDomain,
      databaseURL: config.firebase.databaseURL,
      projectId: config.firebase.projectId,
      storageBucket: config.firebase.storageBucket,
      messagingSenderId: config.firebase.messageSenderId,
      appId: config.firebase.appId,
      measurementId: config.firebase.measurementId,
})
```

This helped me resolve this issue

**Comment by durbanitas** [comment]
Got a same issue. Can't access the public and private keys inside the `/server/api/`-folder:

```js
// nuxt.config.js
export default defineNuxtConfig({
  publicRuntimeConfig: {
    SUPABASE_KEY: process.env.SUPABASE_KEY,
    SUPABASE_URL: process.env.SUPABASE_URL
  },
  privateRuntimeConfig: {
    API_ROUTE_SECRET: process.env.API_ROUTE_SECRET
  },
})
```
same result with the given [example from the documentation](https://v3.nuxtjs.org/guide/features/runtime-config)

```js
// nuxt.config.js
export default defineNuxtConfig({
  runtimeConfig: {
    secretKey: process.env.API_ROUTE_SECRET,
    public: {
        SUPABASE_KEY: process.env.SUPABASE_KEY,
        SUPABASE_URL: process.env.SUPABASE_URL
    }
  },
})
```

```js
// server/api/test.js
export default defineEventHandler(async (e) => {
  const config = useRuntimeConfig();
  console.log(config);
})
```

### output of `console.log(config)`
> {
  app: { baseURL: '/', buildAssetsDir: '/_nuxt/', cdnURL: '' },
  nitro: { routes: {}, envPrefix: 'NUXT_' },
  public: {}
}

An [open ticket in the Nitro repository seems to be aware of the problem](https://github.com/unjs/nitro/issues/246) that the `.env`-variable is not loaded in Nitro.

Any suggestions to access the `.env`-inside the `/server`-folder?

**Reply by danielroe** [reply]
You need to provide a fallback to those variables (or define them when building). https://stackblitz.com/edit/github-gdggof.

**Reply by kang49** [reply]
I just realized that I don't need to put NUXT_ prefix to use it for this. Thanks @danielroe

**Reply by TomMiller-mas** [reply]
Realize it just isn't NUXT_ .  It is NUXT_PUBLIC_ or NUXT_PRIVATE_.  If you don't add these, it won't read at run time.  If you are binding the value at "build", then why have an environment variable?

**Reply by kang49** [reply]
IDK but it work 🤔

`Nuxt3`
## .env:
```
GITHUBKEY=
```

## runtimeConfig:
```
runtimeConfig: {
    GITHUBKEY: process.env.GITHUBKEY
  }
  ```
  
  ## How i'm use (server/api/test.ts):
  ```
  export default defineEventHandler(async (event) => {
    const config = useRuntimeConfig();
    
    const octokit = new Octokit({
            auth: config.GITHUBKEY
        });
  });
  ```

**Reply by TheAlexLichter** [reply]
the better way would be ([as explained in the video](https://www.youtube.com/watch?v=_FYV5WfiWvs)) using the NUXT_ prefix for your env vars:

.env

```
NUXT_GITHUBKEY=
```

config

```ts
runtimeConfig: {
  // can be all caps too
  githubkey: 'YOUR_DEFAULT_HERE'
}
```

Usage:

```ts
export default defineEventHandler(async (event) => {
  const config = useRuntimeConfig();
  
  const octokit = new Octokit({
          auth: config.githubkey
      });
});
```

**Reply by kang49** [reply]
Okay, it's my fault Lol. I forgot that `process.env` can't read ENV with the prefix NUXT_

 <img src="https://media1.tenor.com/m/uiJRHFfY2KUAAAAC/miyano-my-bad.gif" width="200">
 
## Previous runtimeConfig code:
```typescript
runtimeConfig: {
    GITHUBKEY: process.env.GITHUBKEY,
  }
 ```
 
## Current runtimeConfig code:
```typescript
runtimeConfig: {
    GITHUBKEY: process.env.NUXT_GITHUBKEY
  }
  ```
  
  It's work!! Thanks @manniL 💖

**Reply by TheAlexLichter** [reply]
BTW: The right hand side can stay empty (as in my example) as it will only be used as fallback/default value, as long as the env is also present when *running*/*starting* the app

No problem!

**Comment by motau** [comment]
I also encountered this problem after updating from Nuxt3 beta to Nuxt RC4. After build (Nitro preset node_server) my ENV variables were always only available as default values. After some adjustments and unsatisfactory workarounds I found out the solution...

...by reading code ;-)
All systems **Nuxt app ENV Key MUST start with prefix** `NUXT_` or `NUXT_PUBLIC_` for the public scope. 

Example:
* Assume in nuxt.config.ts we have these keys with empty defaults: 
```
      { ...,
        runtimeConfig {
          xyzAuth:'', 
          public: myNoise:'' },
      ...} 
    // here the keys could also be written in the UPPERCASE underscore style, 
    // XYZ_AUTH or MY_NOISE -> but not in your .env (we had used it before this breaking change)
```
* at server start the systems ENV Keys are set (can also be stored in a .env file):
    `$bash NUXT_XYZ_AUTH=breaking NUXT_PUBLIC_MY_NOISE=change pnpm start`
    * in my case the env came from a kubernetes helm chart manifest or local from .env and running pnpm dev

* as we already know the same method for client and server can be used to resolve the values:
  `NUXT_XYZ_AUTH` is dynamically resolved with `useRuntimeConfig().xyzAuth // breaking`
  `NUXT_PUBLIC_MY_NOISE` is dynamically resolved with `useRuntimeConfig().myNoise // change`
 

Relevant code
at `node_modules/.pnpm/@nuxt+schema@3.0.0-rc.4/node_modules/@nuxt/schema/dist/index.d.ts`

```
   ## line 1328 ...
   * @note Environment variables starting with `NUXT_ENV_` are automatically injected
   * into the process environment.
   * 
   * @version 2
  */
  env: {
    [key: string]: any
  },
 ....
```

**Comment by alexcroox** [comment]
how do you configure Nuxt 3 to use a particular env file? I have `.env.local` and `.env.development` files that I like to dynamically load but I don't see any docs around specifying the filename like @nuxt/dotenv module did.

**Reply by mirumirumi** [reply]
I have the same question.

The `.env` files work fine without going through `nuxt.config.ts`, but `.env.dev` and `.env.development` are undefined.
What is the best approach to switch environments?

**Reply by Chippd** [reply]
I was able to achieve this by updating my nuxt.config.ts file to include the code below. 

I have two env files in my root directory, `.production.env` and `.development.env`.

With the below code, it loads development in development and production when running `npm run build`

```
import * as dotenv from 'dotenv'

const path = require('path');
const fs = require('fs');
const env = process.env.NODE_ENV;

const envPath = path.resolve(process.cwd(), `.${env}.env`);
const defaultEnvPath = path.resolve(process.cwd(), '.env');

dotenv.config({
  path: fs.existsSync(envPath) ? envPath : defaultEnvPath,
})
```

**Reply by nirazul** [reply]
Normally, you could use the `--dotenv` arg to set the path to the env file you want. However, this seems to be broken at the moment. 
Based on @Chippd's code, I was able to make it work with any given path (via arg):

```
import minimist from 'minimist'
import * as dotenv from 'dotenv'

const DOT_ENV_ARG = minimist(process.argv.slice(2)).dotenv
const ENV_PATH = resolve(process.cwd(), DOT_ENV_ARG)
const DEFAULT_ENV_PATH = resolve(process.cwd(), '.env')

dotenv.config({
  path: existsSync(ENV_PATH) ? ENV_PATH : DEFAULT_ENV_PATH,
})
```

**Comment by 8area8** [comment]
Hello, I'm using nuxt-bridge with the ssr set to false

And it's so far impossible for me to retrieve an environment variable inside a component, either with vite (import.meta crashes the app), nuxt env (process.env is empty), the vite-plugin-environment plugin ( empty process.env), or runtimeConfig (useRuntimeConfig returns an empty object).

Is it possible to access environment variables on bridge? :(

**Comment by ckhatton** [comment]
I'm trying to call `process.env.npm_package_name` in `nuxt.config.ts` like I used to in Nuxt 2, but it is returning undefined. Anyone have any ideas why it is undefined or where it has been moved to?

Example:
```ts
export default defineNuxtConfig({
  app: {
    head: {
      title: process.env.npm_package_name || '',
      charset: 'utf-8',
      viewport: 'width=device-width, initial-scale=1',
      meta: [
        { name: 'description', content: process.env.npm_package_description || '' },
        { name: 'robots', content: 'noindex' },
        { name: 'googlebot', content: 'noindex' }
      ],
      link: [
        { rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' }
      ]
    }
  }
});
```

**Reply by gavin-hardaker** [reply]
Try doing this:

import.meta.env.npm_package_description

Nuxt now uses vite instead of webpack - so you should be able to get env vars like this now :)

https://vitejs.dev/guide/env-and-mode.html

**Reply by ckhatton** [reply]
That makes sense, but it creates an error...

```fish
ERROR  Cannot restart nuxt:  Cannot use 'import.meta' outside a module                                                     08:58:52

  title: import.meta.env.npm_package_name || '',
  ^^^^
```

I'm surprised they chose `import` as a syntax.

**Reply by gavin-hardaker** [reply]
Hmmmmm yeah, just tried it in one of my other apps, seems like it's related to the public runtime config 

```javascript
export default defineNuxtConfig({
publicRuntimeConfig: {
    NPM_PACKAGE_NAME: process.env. npm_package_name
},
// ...
)}
```

So as a suggestion, 

Why not move your head config to your app.vue file then you can access the config like this:

```javascript
this.$config.NPM_PACKAGE_NAME
```

or 

```html
<script setup>
    const config = useRuntimeConfig();
    const packageName = config.NPM_PACKAGE_NAME
</script>
```

link for reference :) https://serversideup.net/using-environment-variables-in-nuxt-3/

**Reply by ckhatton** [reply]
Born genius. Thank you.

It works for `process.env.npm_package_name`, but not for `process.env.npm_package_description` for some reason. The NPM documentation states that all variables should be there (https://docs.npmjs.com/cli/v6/using-npm/scripts#packagejson-vars). Odd 😅

**Reply by ckhatton** [reply]
```
> const process = require('process')
> process.env
```

Running `process.env` in the debug console doesn't even show `npm_package_name` in the object, let alone `npm_package_description`.

I've opened it up to the floor 😅
https://stackoverflow.com/questions/74111233/nuxt-v3-how-does-one-access-the-package-json-variables

**Reply by gavin-hardaker** [reply]
Okay I think I got it 
![image](https://user-images.githubusercontent.com/10328371/196442508-699b94ad-a244-47a6-ae2b-5137295dc7bd.png)
Try it like this 

```javascript
import process from "process";
process.env.npm_package_description
```

**Reply by ckhatton** [reply]
Like this?

`nuxt.config.ts`
```ts
import process from "process";

export default defineNuxtConfig({
  publicRuntimeConfig: {
    NPM_PACKAGE_NAME: process.env.npm_package_name,
    NPM_PACKAGE_DESCRIPTION: process.env.npm_package_description
  },
  .
  .
  .
});
```

It still doesn't pull through for me. I wonder if it's actually an issue with Nuxt3, that needs reporting?

On that Stack Overflow post, someone kindly pointed out I could just import the `package.json`, but that's an alternative solution, easy way out, and glides over the issue 😄

`nuxt.config.ts`
```ts
import packageJSON from './package.json';

export default defineNuxtConfig({
  app: {
    head: {
      title: packageJSON.name || '',
      charset: 'utf-8',
      viewport: 'width=device-width, initial-scale=1',
      meta: [
        { name: 'description', content: packageJSON.description || '' }
      ]
    }
  },
  .
  .
  .
});

**Reply by gavin-hardaker** [reply]
![image](https://user-images.githubusercontent.com/10328371/196454181-474d1d25-3758-42e0-8c29-92922588ded5.png)

**Reply by gavin-hardaker** [reply]
```
It still doesn't pull through for me. I wonder if it's actually an issue with Nuxt3, that needs reporting?
```

It could be a bug or it could be intentionally like that

**Reply by ckhatton** [reply]
Yeah, I think I'm going to have to let bygones be bygones, and let it go 😏

At least we can be one of the few in the world that know about it 😌

**Comment by Ewocker** [comment]
Is there a way to use nuxt 3 with `ssr: false` and dynamically get the runtimeConfig with `process.env`.
Ex.
I have two environment `dev` and `prod`.
```js
// nuxt.config.ts
{
  ssr: false,
  runtimeConfig: {
    public: {
      ENVIRONMENT: process.env.ENVIRONMENT,
    },
  },
}
```
I found that when setting ssr to false, it always put those value hardcoded into `window.__NUXT__` in `*.html` files.

**Reply by TheAlexLichter** [reply]
Short answer: No

Long answer: If you pre-render all of your HTML, or just a single index.html (SPA-style), your env variables will be "baked in". There is no way to update them then "during runtime" with the runtimeConfig anymore.

**Reply by SPH73** [reply]
If you really don't want to use runtime config then as Nuxt is using Vite you can prefix your variables in your .env file with VITE_  (VITE_SECRET) and use import.meta.env.VITE_SECRET in your app.

**Reply by TomMiller-mas** [reply]
This is still a bug.  It works when you just run the server inside Vs Code.  But as soon as you build it with VITE it doesn't work. This is a pretty serious bug.  I love making a special build for the test server, and then another special build for the staging server, and then another one for production.

**Reply by danielroe** [reply]
The way for runtime configuration is using `runtimeConfig` in Nuxt, and I'm not aware of any issues with runtime config. You should not use `import.meta.env` or the special `VITE_` variables as these are not designed to be updated at runtime.

**Reply by TomMiller-mas** [reply]
I have tried it with no prefix, VITE, NUXT, NUXT_PUBLIC and it doesn't work. I search through the .output directory and every time it is hard coded to the value of the environment variable in the .env file.  So what am I doing wrong?

```
runtimeConfig: {
    public: {
      baseURL:
        process.env.NUXT_PUBLIC_AUTH_SRV_BASE_URL ||
        'https://authapi.mas.systems/auth',
    },
  },
```

```
 runtimeConfig: {
    public: {
      baseURL:
        import.meta.env.VITE_AUTH_SRV_BASE_URL ||
        'https://authapi.mas.systems/auth',
    },
  },

```

**Reply by TheAlexLichter** [reply]
@TomMiller-mas You don't need to set any default value

If you use

```ts
runtimeConfig: {
    public: {
      baseURL: ''
    },
  },
```

 You **have to use** `NUXT_PUBLIC_BASE_URL` to override this specific runtime config variables (`public.baseURL`) during runtime.

**Reply by TomMiller-mas** [reply]
I have more than one server on this box, I can't just use BASE_URL

```
runtimeConfig: {
    public: {
      authSrvBaseURL: ''
    },
  },
```
would that get automatically overridden with NUXT_PUBLIC_AUTH_SRV_BASE_URL ?

**Reply by danielroe** [reply]
yep!

**Reply by TomMiller-mas** [reply]
This is weird, on Windows it works fine where I develop.  I do a build and run the build version and it works fine.  Then I move the build to the Linux test server and I get this error:

`<script>window.__NUXT__={};window.__NUXT__.config={public:{authSrvBaseURL:"http://192.168.1.66:4101",emAppBaseURL:"http://192.168.1.66:4000",i18n:{baseUrl:"",locales:{en:{domain:""},es:{domain:""},fr:{domain:""}}}},app:{baseURL:"/",buildAssetsDir:"/_nuxt/",cdnURL:""}}</script>`

I get this error whether ssr is true or false.  It is picking up the env var properly, but why is it getting put as window.__NUXT variable?  Bug?

**Reply by TheAlexLichter** [reply]
I'm not sure what you mean here exactly: Which exactly is the error you mention 🤔 

It is correct that the public part of the runtimeConfig will be initialized as `__NUXT__.config`

**Reply by TomMiller-mas** [reply]
I found the issue, not sure what is causing it.

**Reply by TheAlexLichter** [reply]
Which issue though? I don't see any error 🤔

**Reply by TomMiller-mas** [reply]
Figured out I had tried too many things at one time and they were canceling each other out.  Using a technique to change the name of the variable I needed to use due to convention.  Primarily because fetch is looking for baseURL

```
  const {
    public: { authSrvBaseURL: baseURL },
  } = useRuntimeConfig();
```

**Comment by mudmixer-wes** [comment]
From [the docs](https://nuxt.com/docs/guide/directory-structure/env#dev-build-and-generate-time), 

```
After your server is built, you are responsible for setting environment variables when you run the server.
Your .env file will not be read at this point. How you do this is different for every environment.
```

Would be super useful if they shared "how you do this... for every environment" since not pulling in a .env file if its running post-build is kind of wild to me.

**Reply by TomMiller-mas** [reply]
The documentation isn't very good.  They really don't emphasize enough that this is done by convention. You don't have to write any code and it will automatically read the environment variables at run time. I got most of this from a forum post for other confused programmers.

Convention:  Prepend either NUXT_PRIVATE or NUXT_PUBLIC to the environment variables and they will automatically be read in.
In code do not include the NUXT_PRIVATE or NUXT_PUBLIC as part of the environment variable in JavaScript Code

NUXT_PUBLIC_DBUSER in the environment file or registry and DBUSER should be in the nuxt.config.ts  Case doesn't seem to matter.

`runtimeConfig: {
    public: {
      dbuser: 'mydbadmin'`

Assign a value here and it becomes the default.  The system automatically picks up the value from the environment variable file or registry with no code based on convention naming at runtime. Let me know if you need more info.

**Comment by ssyberg** [comment]
I know this is super old but I'm still struggling with how to expose env vars at _build time_ (not runtime). I don't see this documented anywhere.

**Reply by TheAlexLichter** [reply]
I still think runtimeConfig + setting env vars during build time is the way.

**Reply by ssyberg** [reply]
Thanks, I was seeing a lot of weirdness but I think all the noise just confused me ultimately and I was trying things that didn't work. Regular old env vars via process.env do appear to be working for me now.

**Reply by TomMiller-mas** [reply]
Have you done a "build" and run that version?  It won't work on a built version.

**Comment by spraddles** [comment]
I love Nuxt, and I use both the front & back ends for full-stack development. But honestly when it comes to ENV vars this one should be really simple, especially on the server side. We should just be able to use the `.ENV` file because that's what it's there for - to store `Environment Variables`.

We should not have to repeat ourselves & redeclare these variables again just for the purpose of getting `Nuxt` to work.

**Reply by danielroe** [reply]
This is absolutely a security decision - otherwise it’s very easy for people to accidentally expose secrets to their end users without realising they have done so.

**Reply by spraddles** [reply]
I get that, but are you saying then that all other stacks (i.e. non-Nuxt) that use the `ENV` file aren't secure? I don't understand sorry...
Also my requirement is for the vars to be used Server-side, not Client-side.
ps: I got this to work by adding the `@nuxtjs/dotenv` package, without needing to add any `runtimeConfig`
